File: ruby/tournament/tournament.rb
=begin
Write your code for the 'Tournament' exercise in this file. Make the tests in
`tournament_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/tournament` directory.
=end
class Row

    attr_accessor :matches_played, :wins, :draws, :losses, :points

    def initialize(matches_played = 0, wins = 0, draws = 0, losses = 0, points = 0)
        @matches_played = matches_played
        @wins = wins
        @draws = draws
        @losses = losses
        @points = points
    end

    def pretty_print(team)
        "%-30s | %2d | %2d | %2d | %2d | %2d\n" % [team, @matches_played, @wins, @draws, @losses, @points]
    end
end

class Tournament
    def self.parse_input(input)
        data = {}
        for row in input.split(/\n/, -1) do
            if row.strip.length > 0
                fields = row.split(/;/, -1)
                team_1 = fields[0]
                team_2 = fields[1]
                status = fields[2]

                if (!data.has_key?(team_1))
                    data[team_1] = Row.new
                end
                
                if (!data.has_key?(team_2))
                    data[team_2] = Row.new
                end

                data[team_1].matches_played += 1
                data[team_2].matches_played += 1
                if (status == "win")
                    data[team_1].wins += 1
                    data[team_1].points += 3
                    data[team_2].losses += 1
                elsif (status == "loss")
                    data[team_1].losses += 1
                    data[team_2].wins += 1
                    data[team_2].points += 3
                else # draw
                    data[team_1].draws += 1
                    data[team_1].points += 1
                    data[team_2].draws += 1
                    data[team_2].points += 1
                end
            end
        end
        data
    end

    def self.tally(input)
        header = "Team                           | MP |  W |  D |  L |  P\n"
        data = parse_input(input)
        ret = header
        for key, value in data.sort_by { |key, value| [-value.points, key] } do
            ret += value.pretty_print(key)
        end
        ret
    end
end

File: javascript/tournament/tournament.js
//
// This is only a SKELETON file for the 'Tournament' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const header = 'Team                           | MP |  W |  D |  L |  P'

export const tournamentTally = (data) => {
  let rows = data.split("\n").map((row) => row.split(";"))
  let tally = {}
  if (data.length > 0)
  {
    for(let row of rows) 
    {
      let [team1, team2, state] = row
      addTeam(tally, team1)
      addTeam(tally, team2)

      tally[team1]["MP"] += 1
      tally[team2]["MP"] += 1

      switch(state)
      {
        case 'win':
          tally[team1]["W"] += 1
          tally[team1]["P"] += 3

          tally[team2]["L"] += 1
          break;
        case "loss":
          tally[team1]["L"] += 1

          tally[team2]["W"] += 1
          tally[team2]["P"] += 3
          break;
        case "draw":
          tally[team1]["D"] += 1
          tally[team1]["P"] += 1

          tally[team2]["D"] += 1
          tally[team2]["P"] += 1
          break;
        default:
          throw new Error("Invalid game state")
      }
    }
  }
  let computed = []
  if (Object.keys(tally).length > 0)
    computed = [... Object.values(tally)]
  computed.sort((a, b) => a["P"] === b["P"] ? a["Team"].localeCompare(b["Team"]) : b["P"] - a["P"])


  return formatResults(computed)
};

const formatResults = (data) => {
  let ret = header
  for (let row of data)
  {
    ret += `\n${row.Team.padEnd(30)} | ${String(row.MP).padStart(2)} | ${String(row.W).padStart(2)} | ${String(row.D).padStart(2)} | ${String(row.L).padStart(2)} | ${String(row.P).padStart(2)}`
  }
  return ret
}

const addTeam = (data, teamName) => {
  if (!(teamName in data))
  {
    data[teamName] = {"Team": teamName, "MP": 0, "W": 0, "D": 0, "L": 0, "P": 0}
  }
}
